flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/DynamicKafkaRecordSerializationSchema.java

```diff
@@ -144,14 +147,34 @@ public void open(
         valueSerialization.open(context);
     }
 
+    private String getTargetTopic(RowData element) {
+        final String topic = readMetadata(element, KafkaDynamicSink.WritableMetadata.TOPIC);
+        if (topic == null && topics == null) {
+            throw new IllegalArgumentException(
+                    "The topic of the sink record is not valid. Expected a single topic but no topic is set.");
+        } else if (topic == null && topics.size() == 1) {
+            return topics.get(0);
```

Comment:
Can we start wit the `topics.size == 1` check and immediately return if the connector is used with a static target topic? This could save the processing in `readMetadata`.

Reply:
Good idea, I thought about it and thought it might make the semantics a bit complicated, because if we do this, then for `topics.size == 1` then I think the `topic` metadata column should not be writable, but I think it's worth saving the processing. Updated

Reply:
> then for topics.size == 1 the topic metadata column should not be writable

I think that makes a lot of sense!